src/app/logproc/hashfarm.ml

```diff
@@ -0,0 +1,51 @@
+open Core
+
+let all_pairs (xs : 'a list) (ys : 'b list) : ('a * 'b) list =
+  List.bind ~f:(fun x -> List.map ~f:(fun y -> (x, y)) ys) xs
+
+let all_names =
+  lazy
+    ( all_pairs Hashfarm_wordlists.adjectives
+        (all_pairs Hashfarm_wordlists.colors Hashfarm_wordlists.animals)
+    |> Array.of_list
+    |> Array.map ~f:(fun (a, (b, c)) -> a ^ " " ^ b ^ " " ^ c) )
+
+let used_names : (string, string) Hashtbl.t = Hashtbl.create (module String)
+
+let select_name x =
+  let all_names = force all_names in
+  let name_for_hash =
+    all_names.((String.hash x :> int) % Array.length all_names)
+  in
+  match Hashtbl.find used_names name_for_hash with
+  | Some preimage ->
+      if not (String.equal preimage x) then (
+        eprintf "FATAL:hashfarm collision\n" ;
+        exit 1 )
+      else name_for_hash
+  | None ->
+      Hashtbl.add_exn used_names ~key:name_for_hash ~data:x ;
+      name_for_hash
+
+let rec map_data_hashes (sm : Yojson.Safe.json) ~f : Yojson.Safe.json =
+  let map_data_hashes = map_data_hashes ~f in
+  match sm with
+  | `Null | `Bool _ | `Int _ | `Intlit _ | `Float _ | `String _ -> sm
+  | `Assoc [("logproc_interp", `String "data_hash"); ("value", `String hash)]
```

Comment:
It would be better if we have a function that looks if an association list has these fields, rather than explicitly having two similar patterns.